rpm/dnf.go

```diff
@@ -0,0 +1,141 @@
+package rpm
+
+import (
+	"context"
+	"database/sql"
+	_ "embed" // embed query
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+
+	"github.com/quay/zlog"
+	_ "modernc.org/sqlite" // register the sqlite driver
+)
+
+// BUG(hank) The dnf mapping is less useful than it could be because there's no
+// reliable way to turn the "repoid" that it reports into something with meaning
+// outside of the Red Hat build system's builder's context. See [CLAIRDEV-45]
+// for more information.
+//
+// [CLAIRDEV-45]: https://issues.redhat.com/browse/CLAIRDEV-45
+
+// RepoMap reports the latest nevra → repoid mapping, as extracted from the dnf
+// or dnf5 database. If the layer is known to have nonsense repoids, nothing is
+// reported.
+func repoMap(ctx context.Context, sys fs.FS) (map[string]string, error) {
+	var toOpen string
+	var isdnf5 bool
+Look:
+	for i, p := range []string{
+		`usr/lib/sysimage/libdnf5/transaction_history.sqlite`,
+		`var/lib/dnf/history.sqlite`,
+	} {
+		switch _, err := fs.Stat(sys, p); {
+		case errors.Is(err, nil):
+			toOpen = p
+			isdnf5 = i == 0
+			break Look
+		case errors.Is(err, fs.ErrNotExist): // OK
+		default:
+			return nil, fmt.Errorf("rpm: unexpected error opening dnf history: %w", err)
+		}
+	}
+	if toOpen == "" {
+		// Nothing found.
+		return nil, nil
+	}
+	if fi, err := fs.Stat(sys, `root/buildinfo/content_manifests`); errors.Is(err, nil) && fi.IsDir() {
+		// This is a RHEL layer, skip.
+		return nil, nil
+	}
```

Comment:
- [ ] Update this to check the files for a marker value and continue if found.

Should look for `from_dnf_hint` key in the yaml files.
See also: containerbuildsystem/atomic-reactor#2140